packages/ra-core/src/controller/input/useReferenceArrayInputController.ts

```diff
@@ -105,7 +105,10 @@ const useReferenceArrayInputController = ({
         : referenceRecordsFromStore;
 
     // filter out not found references - happens when the dataProvider doesn't guarantee referential integrity
-    const finalReferenceRecords = referenceRecords.filter(Boolean);
+    // filter out not selected items - happens when the form value is changed by script (i.e. with a filtering Button)
+    const finalReferenceRecords = referenceRecords
+        .filter(Boolean)
+        .filter(item => input.value.includes(item.id));
```

Comment:
I think the problem isn't there, it's in the useEffect line 58 that does not properly work in all cases.